expr/plan.go

```diff
@@ -142,49 +175,106 @@ func consumeKwarg(optArgs []optArg, namedArgs map[string]*expr, k string, seenKw
+	case argFloat:
+		switch got.etype {
+		case etInt:
+			// integer is also a valid float, just happened to have no decimals
+			*v.val = float64(got.int)
+		case etFloat:
+			*v.val = got.float
+		default:
+			return ErrBadKwarg{k, exp, got.etype}
 		}
-	case str:
+	case argString:
 		if got.etype != etString {
-			return ErrBadKwarg{k, str, got.etype}
+			return ErrBadKwarg{k, exp, got.etype}
 		}
+		*v.val = got.str
 	}
 	return nil
 }
 
+func consumeSeriesArg(args []*expr, j int, exp arg, from, to uint32, stable bool, reqs []Req) (int, []Req, error) {
+	got := args[j]
+	var err error
+	var fn Func
+	switch v := exp.(type) {
+	case argSeries:
+		if got.etype != etName && got.etype != etFunc {
+			return 0, nil, ErrBadArgumentStr{"func or name", string(got.etype)}
+		}
+		fn, reqs, err = newplan(got, from, to, stable, reqs)
+		if err != nil {
+			return 0, nil, err
+		}
+		*v.val = fn
+	case argSeriesList:
+		if got.etype != etName && got.etype != etFunc {
+			return 0, nil, ErrBadArgumentStr{"func or name", string(got.etype)}
+		}
+		fn, reqs, err = newplan(got, from, to, stable, reqs)
+		if err != nil {
+			return 0, nil, err
+		}
+		*v.val = fn
+	case argSeriesLists:
+		if got.etype != etName && got.etype != etFunc {
+			return 0, nil, ErrBadArgumentStr{"func or name", string(got.etype)}
+		}
+		fn, reqs, err = newplan(got, from, to, stable, reqs)
+		if err != nil {
+			return 0, nil, err
+		}
+		*v.val = append(*v.val, fn)
+		// special case! consume all subsequent args (if any) in args that will also yield a seriesList
+		for len(args) > j+1 && (args[j+1].etype == etName || args[j+1].etype == etFunc) {
+			j += 1
+			fn, reqs, err = newplan(args[j], from, to, stable, reqs)
+			if err != nil {
+				return 0, nil, err
+			}
+			*v.val = append(*v.val, fn)
+		}
+	default:
+		panic("unsupported type for consumeSeriesArg")
+	}
+	j += 1
+	return j, reqs, nil
+}
+
 // newplan adds requests as needed for the given expr, resolving function calls as needed
-func newplan(e *expr, from, to uint32, stable bool, reqs []Req) ([]Req, error) {
+func newplan(e *expr, from, to uint32, stable bool, reqs []Req) (Func, []Req, error) {
 	if e.etype != etFunc && e.etype != etName {
-		return nil, errors.New("request must be a function call or metric pattern")
+		return nil, nil, errors.New("request must be a function call or metric pattern")
 	}
 	if e.etype == etName {
-		reqs = append(reqs, Req{
+		req := Req{
 			e.str,
 			from,
 			to,
-		})
-		return reqs, nil
+		}
+		reqs = append(reqs, req)
+		return NewGet(req), reqs, nil
 	}
 
 	// here e.type is guaranteed to be etFunc
 	fdef, ok := funcs[e.str]
 	if !ok {
-		return nil, ErrUnknownFunction(e.str)
+		return nil, nil, ErrUnknownFunction(e.str)
 	}
 	if stable && !fdef.stable {
-		return nil, ErrUnknownFunction(e.str)
+		return nil, nil, ErrUnknownFunction(e.str)
 	}
 
 	fn := fdef.constr()
-	return newplanFunc(e, fn, from, to, stable, reqs)
+	reqs, err := newplanFunc(e, fn, from, to, stable, reqs)
```

Comment:
that's actually how it used to be. i split it off into a separate function to separate the logic dealing with A) expr is a name vs a func, and if func, look it up and plan it. B) create the plan for a given func.
I think this is a reasonable separation, and the main reason I wanted it is that so that we can call newplanFunc with any given function, which we do to test planning of a function call with optional args (independently of function lookup).